stdlib: remove some static assertions which may not hold

compnerd opened the pull request:
These are based around the idea that `std::atomic` is trivially
constructible, which is not a guarantee that the standard fully
provides.  The default initialization of the `std::atomic` type
may leave it in an undetermined state.  These were caught using
the Visual C++ preview runtime.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
@lorentey @moiseev @gparker42

lorentey commented:
cc @mikeash

lorentey commented on stdlib/public/SwiftShims/GlobalObjects.h:
```diff
@@ -102,13 +102,6 @@ static_assert(
-              "empty dictionary type should be POD");
-static_assert(std::is_pod<_SwiftEmptySetSingleton>::value,
-              "empty set type should be POD");
-
```
IIUC, the issue is caused by `HeapObject`, so we could (and probably should) still check for the PODness of `_Swift{Array,Dictionary,Set}BodyStorage` here.

compnerd replied:
Yes, the issue was caused by `HeapObject`.  However, the singletons contain a `HeapObject` type, and that fails the PODness check  :-(.

lorentey replied:
Sure, but the `*BodyStorage` types are still PODs. (Checking those is of limited, but probably nonzero, benefit.)

compnerd replied:
@lorentey - hmm, interesting, since I definitely am seeing those checks fail as well.  I can try to dig through those and split this up into two sets, one to remove the other assertions, and one to deal with the PODness here.

lorentey replied:
Oh, that's weird! They're simple structs of integers and pointers with no explicit constructors/destructors. Are typedefs like `__swift_intptr_t` not considered POD on MSVC?

The stdlib's Swift code also assumes these are trivial types -- for example, it doesn't distinguish between initialization and assignment when creating such objects.

compnerd replied:
I'm almost 100% certain that it is what I said earlier:

```
struct _SwiftEmptyDictionarySingleton {
  struct HeapObject header;
  struct _SwiftDictionaryBodyStorage body;
  __swift_uintptr_t metadata;
};
```

Note that the structure itself contains an **inline** copy of `HeapObject`.  That breaks the PODness of the type, because the `HeapObject` is no longer a POD type due to the `std::atomic<uint64_t>` in it breaking that.

lorentey replied:
I'm suggesting to check `*BodyStorage` for podness, not `*Singleton`.

compnerd replied:
Bleh, apparently, I can't read :-(.  The removed assertions here are for the Singletons, and the `_SwiftEmptyArrayStorage` which has an inline `HeapObject` header.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Linux platform

jrose-apple commented:
I don't think we *can* remove these assertions. We rely on a HeapObject coming out of calloc ready to go.

compnerd commented:
@jrose-apple - that is not a guarantee that the C++ standard makes.  The `std::atomic` needs to be initialized.  This was the discussions that @lorentey and I had above, and he also seemed to agree that this was the correct thing to do.

mikeash commented:
From the random copy of C++11 that turned up in a search:

> The `atomic` integral specializations and the specialization `atomic<bool>` shall have standard layout. They shall each have a trivial default constructor and a trivial destructor.

Which I think means it has to have the same representation as the integer type, and not do any initialization. I'm bad at standardese though so draw your own conclusions.

compnerd commented:
Yes, it has the same representation, but the constructor is needed to initialize it to 0 (otherwise it is in an indeterminate state).  The zero-ing of the member is sufficient to make it fail the trivial constructability test.  See LWG Issue 2334.  LWG P0883 is designed to address this which will make the `std::is_trivially_constructible` no longer hold.

jrose-apple commented:
So we can still put them in constant sections because of the newly-added `constexpr`, but we can't just calloc a HeapObject or HeapObject-containing thing anymore. Maybe that's okay if we actually rewrite the stdlib to do that, but let's do that first and *then* remove the static_asserts rather than the other way around.

compnerd commented:
@jrose-apple - hmm, perhaps you can file a radar for that?  I think that this would be something that Apple might want to track more closely (but, this was found by building with the next version of the C++ runtime from Microsoft which does actually implement parts of the future specification), so this will become a problem for everyone.

jrose-apple commented:
@mikeash, what do you think?

mikeash commented:
It looks like we don't actually rely on the memory being zero-filled at allocation as I was thinking. We call through to `malloc`, not `calloc`, and then initialize the header manually by setting the metadata pointer to the metadata, and setting the refcounts to whatever `RefCountBits(0, 1)` ends up being.

That initialization has an interesting comment:

    // FIXME: this should be a placement new but that adds a null check
    object->metadata = metadata;
    object->refCounts.init();

Can we find a way to use placement new without penalty? Seems like that would be the best for everything.

compnerd commented:
@mikeash - yes and no.  It is "removable" by switching to C++17.  As of C++17, a null pointer to the placement new is UB, which means that the `nullptr` check should not be emitted by the compiler.

compnerd commented:
@mikeash - until we can move to C++17, what do you think is the best way to move forward here?

compnerd commented:
@mikeash - ping

mikeash commented:
Sorry to leave you hanging! I don't want to introduce an unnecessary NULL check. If placement new means we have to have the check for now, can we use placement new when compiling with VC++ and keep the existing version for other compilers?

Alternately, is it possible to avoid the NULL check with an annotation to inform the compiler that the pointer is definitely not ever NULL?

Either one would be good by me.

compnerd commented:
I'm happy with a C++17/Windows path through this codepath.  Interestingly enough, I do not see a null pointer check even with C++14:

```
	.def	 "?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z";
	.scl	3;
	.type	32;
	.endef
	.section	.text,"xr",one_only,"?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z"
	.p2align	4, 0x90         # -- Begin function ?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z
"?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z": # @"?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z"
.seh_proc "?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z"
# %bb.0:                                # %entry
	push	rsi
	.seh_pushreg 6
	sub	rsp, 32
	.seh_stackalloc 32
	.seh_endprologue
	mov	rsi, rcx
	mov	rcx, rdx
	mov	rdx, r8
	call	swift_slowAlloc
	mov	qword ptr [rax], rsi
	mov	qword ptr [rax + 8], 2
	add	rsp, 32
	pop	rsi
	ret
	.seh_handlerdata
	.section	.text,"xr",one_only,"?_swift_allocObject_@@YAPEAUHeapObject@swift@@PEBU?$TargetHeapMetadata@UInProcess@swift@@@2@_K1@Z"
	.seh_endproc
```

mikeash commented:
Maybe the compiler has already figured out that it can't be NULL? Let me know and send me a diff if you'd like me to try with clang and see if it produces good code.

compnerd committed:
stdlib: use placement new on Windows, disable asserts

The assertions here are based around the idea that `std::atomic` is
trivially constructible which is not a guarantee that the standard fully
provides.  The default initialization of the `std::atomic` type may
leave it in an undetermined state.  These were caught using the Visual
C++ preview runtime.

Ideally, the object constructor would use a placement new operator.
However, prior to C++17, the C++ standard mandated that there be a
NULL pointer check in the placement new operator.  This is something
which is no longer the case with C++17.  Switch to the placement new
operator for C++17 and newer and enable that codepath for Windows as
well (which seemingly elides the null-pointer check with clang-cl).

compnerd commented:
@mikeash - uploaded a new version of the change that enables it under C++17 and Windows, I'll double check this on Linux as well.

compnerd commented:
@swift-ci please test

mikeash commented:
I'll try the placement new path on Mac and see how it does. If we can be confident that it'll generate good code then maybe we can just use that everywhere.

compnerd commented:
@mikeash - please do, since I just ran this on Linux (with some "hacks" - namely, added `-O2 -g0 -S -o - - -DNDEBUG -U_DEBUG -USWIFT_RUNTIME_ENABLE_LEAK_CHECKER -USWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS` and get this:

```
	.p2align	4, 0x90         # -- Begin function _ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm
	.type	_ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm,@function
_ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm: # @_ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm
	.cfi_startproc
# %bb.0:                                # %entry
	pushq	%rbx
	.cfi_def_cfa_offset 16
	.cfi_offset %rbx, -16
	movq	%rdi, %rbx
	movq	%rsi, %rdi
	movq	%rdx, %rsi
	callq	swift_slowAlloc
	movq	%rbx, (%rax)
	movq	$2, 8(%rax)
	popq	%rbx
	.cfi_def_cfa_offset 8
	retq
.Lfunc_end1:
	.size	_ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm, .Lfunc_end1-_ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm
	.cfi_endproc
```

mikeash commented:
Here's what I get with placement new:

```__ZL19_swift_allocObject_PKN5swift18TargetHeapMetadataINS_9InProcessEEEmm:
000000000033db80	pushq	%rbp
000000000033db81	movq	%rsp, %rbp
000000000033db84	pushq	%rbx
000000000033db85	pushq	%rax
000000000033db86	movq	%rdi, %rbx
000000000033db89	movq	%rsi, %rdi
000000000033db8c	movq	%rdx, %rsi
000000000033db8f	callq	_swift_slowAlloc
000000000033db94	movq	%rbx, (%rax)
000000000033db97	movq	$0x2, 0x8(%rax)
000000000033db9f	addq	$0x8, %rsp
000000000033dba3	popq	%rbx
000000000033dba4	popq	%rbp
000000000033dba5	retq
```

Looks great! Seems like we might as well use it everywhere now. I guess that comment must be obsolete.

compnerd commented:
Awesome!  I'll upload a new version with that enabled globally!

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

mikeash commented:
Woohoo!

mikeash commented:
Go for it!
